Add wrapped text support

twixuss opened the pull request:
![output](https://github.com/user-attachments/assets/89390b6c-96cf-4d70-bd55-bda5b232cb5c)

Wrapping is disabled by default, can be enabled by providing a bool flag in constructor.

It's a bit slow though. 20k line text renders at ~30 fps and when selecting it drops to ~13.

twixuss committed:
Add wrapped text support

AidanSun05 commented on textselect.cpp:
```diff
@@ -150,11 +183,53 @@ void TextSelect::handleMouseDown(const ImVec2& cursorPosStart) {
+        currentLine = getLineAtIdx(y);
+
+        // Calculate index of the sub-line in the current line.
+        std::size_t localWrappedY = static_cast<std::size_t>(std::min(std::max(std::floor((mousePos.y - accumulatedHeight) / textHeight), 0.0f), std::round(currentLineSize.y / textHeight) - 1));
```
The library uses C++17 so we can use `std::clamp()` over the `std::min(std::max(...))` pattern to improve readability.

```suggestion
        std::size_t localWrappedY = static_cast<std::size_t>(std::clamp(std::floor((mousePos.y - accumulatedHeight) / textHeight), 0.0f, std::round(currentLineSize.y / textHeight) - 1));
```

AidanSun05 commented on textselect.cpp:
```diff
@@ -150,11 +183,53 @@ void TextSelect::handleMouseDown(const ImVec2& cursorPosStart) {
+        x = currentSubLine.data() - currentLine.data() + getCharIndex(currentSubLine, mousePos.x);
+    } else {
+        // Get Y position of mouse cursor, in terms of line number (clamped to the valid range)
+        y = static_cast<std::size_t>(std::min(std::max(std::floor(mousePos.y / textHeight), 0.0f), static_cast<float>(numLines - 1)));
```
This is the same as above:

```suggestion
        y = static_cast<std::size_t>(std::clamp(std::floor(mousePos.y / textHeight), 0.0f, static_cast<float>(numLines - 1)));
```

AidanSun05 commented on textselect.cpp:
```diff
@@ -150,11 +183,53 @@ void TextSelect::handleMouseDown(const ImVec2& cursorPosStart) {
+            accumulatedHeight += currentLineSize.y + ImGui::GetCurrentContext()->Style.ItemSpacing.y;
+        }
+        if (y >= numLines)
+            return;
```
Please ensure the bodies of single-line `if` statements are surrounded by braces.

```suggestion
        if (y >= numLines) {
            return;
        }
```

AidanSun05 commented on textselect.cpp:
```diff
@@ -150,11 +183,53 @@ void TextSelect::handleMouseDown(const ImVec2& cursorPosStart) {
+    } else {
+        // Get Y position of mouse cursor, in terms of line number (clamped to the valid range)
+        y = static_cast<std::size_t>(std::min(std::max(std::floor(mousePos.y / textHeight), 0.0f), static_cast<float>(numLines - 1)));
+        if (y >= numLines)
+            return;
```
This is another place where we can add braces.

```suggestion
        if (y >= numLines) {
            return;
        }
```

AidanSun05 commented on textselect.hpp:
```diff
@@ -47,6 +47,8 @@ class TextSelect {
     // initiated by clicking the text, or different element.
     bool shouldHandleMouseDown = false;
+    
+    bool enableWordWrap = true;
```
We set this field in the constructor initializer list so we won't need a dedicated initializer here. Also, a brief comment about what this field does would be nice for consistency.

```suggestion
    // Indicates whether text selection with word wrapping should be enabled.
    bool enableWordWrap;

```

AidanSun05 commented on textselect.cpp:
```diff
@@ -140,6 +140,39 @@ TextSelect::Selection TextSelect::getSelection() const {
+        wrappedLineEnd = font->CalcWordWrapPositionA(1, wrappedLineStart, textEnd, wrapWidth);
+
+        if (wrappedLineEnd - wrappedLineStart != 0) {
+            result.push_back(std::string_view(wrappedLineStart, wrappedLineEnd));
```
We can't use the (start, end) constructor since it was introduced in C++20 and this library uses C++17. Also, using brace-initialization and an explicit `std::size_t` cast matches the rest of the code.

```suggestion
            result.push_back({ wrappedLineStart, static_cast<std::size_t>(wrappedLineEnd - wrappedLineStart) });
```

AidanSun05 reviewed (changes requested):
I left some minor suggestions related to code style which could benefit readability.

AidanSun05 commented:
Thanks for implementing this large feature! It's coming along well, but sometimes drawing the rectangle with wrapped text and UTF-8 characters behaves unexpectedly even though the text copies correctly. This recording below shows triple-click and dragging behavior:

![recording](https://github.com/user-attachments/assets/d3cc5921-26fc-4e61-b8bf-f2726de851c1)

I see a lot of direct manipulation of string indices through pointer arithmetic which would be a likely cause for this issue, since in UTF-8 one byte is not necessarily one character. The preferred way to accomplish this would be to use the functions provided by utfcpp.

Also, thanks for leaving a comment about the performance of this feature. I think the slowness with a lot of text is acceptable for this library, and part of it is caused by Dear ImGui getting pretty slow when it tries to render it all - this is what `ImGuiListClipper` tries to solve, but it's designed for constant-height text so it doesn't work with word wrapping.

twixuss committed:
Apply suggestions from code review

Use c++17 `string_view` constructor, use `clamp` instead of `min`/`max`, brace style changes

Co-authored-by: Aidan Sun <aidansun05@gmail.com>

twixuss committed:
Remove redundant initializer, add comment

Co-authored-by: Aidan Sun <aidansun05@gmail.com>

twixuss committed:
use utf8::distance instead of pointer arithmetic

twixuss committed:
Fix wrong behavior on empty lines.

twixuss committed:
Factor out selection rect drawing

twixuss committed:
missing `clamp`

AidanSun05 reviewed (approved):
Looks great overall. I'll go ahead and merge this now.

One thing that stood out is that selection and rendering now have two separate code paths for wrapped and non-wrapped text. In the future, it might be a good idea for us to consider a small refactor where selection and rendering operate on unified accessor functions which serve as an abstraction over wrapped/non-wrapped text. For example, given the screenshot below:

![Screenshot 2025-04-20 215306](https://github.com/user-attachments/assets/4d45a702-5dfa-4b09-8c7f-5a15cb070cce)

The physical lines (as they appear in memory) are:

```cpp
{
    "Line 1",
    "Line 2",
    "Line 3",
    "A longer line",
    "Text selection in Dear ImGui",
    "UTF-8 characters Ë ⑤ 三【 】┌──┐"
}
```

While the virtual lines (as they appear wrapped on screen) are:

```cpp
{
    "Line 1",
    "Line 2",
    "Line 3",
    "A longer line",
    "Text selection in ", // (note the trailing space)
    "Dear ImGui",
    "UTF-8 characters Ë ",
    "⑤ 三【 】┌──┐"
}
```

Introducing accessor functions like `getVirtualLineAtIdx` and `getVirtualNumLines` could allow us to unify access to physical and virtual lines by abstracting the difference when word wrapping is enabled, while they would behave the same as their current counterparts if word wrapping is disabled. This allows `handleMouseDown` and `drawSelection` to operate purely on virtual lines, which reduces duplication and improves separation of concerns.

One tricky aspect would be handling triple-click and copying, which would require us to join multiple virtual lines together without newlines.

Let me know if you're willing to open a follow-up PR for this refactoring or if you want to discuss further. Thanks for your contributions!

twixuss commented:
Thanks for review.
